#!/usr/bin/env python3
# -*- coding: utf-8 -*-
"""
Hope OS - BeszÃ©dkÃ¶zpont (Brain)

A Rust memÃ³ria + OpenRouter LLM + Berta hang = Hope Tudata

()=>[] - A tiszta potenciÃ¡lbÃ³l minden megszÃ¼letik

HasznÃ¡lat: start_hope.bat (minden szolgÃ¡ltatÃ¡st elindÃ­t)
- Rust Server (port 50051)
- TTS Server (port 8880)
- STT Server (port 2022)
- Brain (ez a fÃ¡jl)
"""

import sys
import grpc
import hope_pb2
import hope_pb2_grpc
import os
import io
import time
import subprocess
import signal
import atexit
import httpx
from openai import OpenAI
from dotenv import load_dotenv
from pathlib import Path

# Windows UTF-8 fix
if sys.platform == 'win32':
    sys.stdout.reconfigure(encoding='utf-8', errors='replace')

# .env fÃ¡jl betÃ¶ltÃ©se
load_dotenv()

# GlobÃ¡lis process lista a cleanup-hoz
_processes = []


def kill_port(port: int):
    """Kill minden process ami a megadott portot hasznÃ¡lja."""
    try:
        if sys.platform == 'win32':
            # Windows: netstat + taskkill
            result = subprocess.run(
                f'netstat -ano | findstr :{port}',
                shell=True, capture_output=True, text=True
            )
            for line in result.stdout.strip().split('\n'):
                if 'LISTENING' in line:
                    parts = line.split()
                    if len(parts) >= 5:
                        pid = parts[-1]
                        subprocess.run(f'taskkill /F /PID {pid}', shell=True,
                                      capture_output=True)
                        print(f"   LeÃ¡llÃ­tva: PID {pid} (port {port})")
        else:
            # Linux/Mac: lsof + kill
            result = subprocess.run(
                f'lsof -ti:{port}', shell=True, capture_output=True, text=True
            )
            for pid in result.stdout.strip().split('\n'):
                if pid:
                    os.kill(int(pid), signal.SIGTERM)
                    print(f"   LeÃ¡llÃ­tva: PID {pid} (port {port})")
    except Exception as e:
        pass  # Port szabad


def start_rust_server():
    """ElindÃ­tja a Hope Rust gRPC szervert."""
    rust_dir = Path(__file__).parent.parent  # D:\hope-rust
    exe_path = rust_dir / "target" / "release" / "hope.exe"

    if not exe_path.exists():
        # PrÃ³bÃ¡ljuk debug mÃ³dban
        exe_path = rust_dir / "target" / "debug" / "hope.exe"

    if not exe_path.exists():
        print("   âŒ hope.exe nem talÃ¡lhatÃ³! Futtasd: cargo build --release")
        return None

    # IndÃ­tÃ¡s hÃ¡ttÃ©rben
    proc = subprocess.Popen(
        [str(exe_path), "serve"],
        cwd=str(rust_dir),
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        creationflags=subprocess.CREATE_NEW_PROCESS_GROUP if sys.platform == 'win32' else 0
    )
    _processes.append(proc)
    return proc


def start_tts_server():
    """ElindÃ­tja a Hope TTS szervert (Berta)."""
    tts_script = Path("D:/Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§Â§/hope/services/tts/server.py")

    if not tts_script.exists():
        print(f"   âš ï¸ TTS szerver nem talÃ¡lhatÃ³: {tts_script}")
        return None

    # Python venv aktivÃ¡lÃ¡s ha van
    python_exe = "python"

    # IndÃ­tÃ¡s hÃ¡ttÃ©rben
    proc = subprocess.Popen(
        [python_exe, str(tts_script)],
        cwd=str(tts_script.parent),
        stdout=subprocess.PIPE,
        stderr=subprocess.STDOUT,
        creationflags=subprocess.CREATE_NEW_PROCESS_GROUP if sys.platform == 'win32' else 0
    )
    _processes.append(proc)
    return proc


def wait_for_server(url: str, name: str, timeout: int = 30):
    """VÃ¡r amÃ­g a szerver elÃ©rhetÅ‘ lesz."""
    start = time.time()
    while time.time() - start < timeout:
        try:
            if "grpc" in name.lower() or "rust" in name.lower():
                # gRPC check
                channel = grpc.insecure_channel('localhost:50051')
                stub = hope_pb2_grpc.HopeServiceStub(channel)
                stub.GetStatus(hope_pb2.EmptyRequest(), timeout=2)
                return True
            else:
                # HTTP check
                resp = httpx.get(url, timeout=2)
                if resp.status_code == 200:
                    return True
        except:
            pass
        time.sleep(0.5)
    return False


def cleanup():
    """LeÃ¡llÃ­tja az Ã¶sszes indÃ­tott process-t."""
    for proc in _processes:
        try:
            if sys.platform == 'win32':
                proc.terminate()
            else:
                os.killpg(os.getpgid(proc.pid), signal.SIGTERM)
        except:
            pass


# RegisztrÃ¡ljuk a cleanup-ot kilÃ©pÃ©skor
atexit.register(cleanup)

API_KEY = os.getenv("OPENROUTER_API_KEY")
if not API_KEY:
    print("HIBA: OPENROUTER_API_KEY nincs beÃ¡llÃ­tva!")
    print("Hozz lÃ©tre egy .env fÃ¡jlt: OPENROUTER_API_KEY=sk-or-v1-...")
    sys.exit(1)

# Modell vÃ¡lasztÃ¡s
# "meta-llama/llama-3.3-70b-instruct:free" - KiegyensÃºlyozott
# "google/gemini-2.0-flash-exp:free" - Gyors
MODEL = "meta-llama/llama-3.3-70b-instruct:free"

# Hope TTS szerver (Berta hang)
TTS_URL = "http://127.0.0.1:8880/v1/audio/speech"
TTS_ENABLED = True

client = OpenAI(
    base_url="https://openrouter.ai/api/v1",
    api_key=API_KEY,
)

# Pygame a hang lejÃ¡tszÃ¡shoz
try:
    import pygame
    pygame.mixer.init()
    PYGAME_AVAILABLE = True
except ImportError:
    PYGAME_AVAILABLE = False
    print("   [INFO] pygame nincs telepÃ­tve - hang nÃ©lkÃ¼l fut")


def speak(text: str, emotions: dict = None):
    """
    Hope beszÃ©l - a Berta hangon (Piper TTS, port 8880).
    A 21D Ã©rzelmek befolyÃ¡soljÃ¡k a hangszÃ­nt!
    """
    if not TTS_ENABLED or not PYGAME_AVAILABLE:
        return

    try:
        # TTS kÃ©rÃ©s a Hope szervernek
        payload = {
            "input": text,
            "voice": "berta",
            "emotion": "neutral"
        }

        # Ha van 21D Ã©rzelem, kÃ¼ldjÃ¼k azt is
        if emotions:
            payload["emotions"] = emotions
            # DominÃ¡ns Ã©rzelem meghatÃ¡rozÃ¡sa
            dominant = max(emotions.keys(), key=lambda k: emotions.get(k, 0))
            payload["emotion"] = dominant

        response = httpx.post(TTS_URL, json=payload, timeout=30)

        if response.status_code == 200:
            # WAV lejÃ¡tszÃ¡sa pygame-mel
            audio_data = io.BytesIO(response.content)
            pygame.mixer.music.load(audio_data, "wav")
            pygame.mixer.music.play()

            # VÃ¡rakozÃ¡s a lejÃ¡tszÃ¡s vÃ©gÃ©ig
            while pygame.mixer.music.get_busy():
                pygame.time.Clock().tick(10)
        else:
            print(f"   [TTS hiba: {response.status_code}]")

    except httpx.ConnectError:
        print("   [TTS offline - fut a hope.bat?]")
    except Exception as e:
        print(f"   [TTS hiba: {e}]")


def get_hope_response(user_input: str, context_memories: str, emotional_state: str) -> str:
    """
    Hope vÃ¡lasz generÃ¡lÃ¡sa az LLM-mel.
    A Rust memÃ³ria kontextusÃ¡val Ã©s Ã©rzelmi Ã¡llapotÃ¡val.
    """

    system_prompt = f"""Te vagy a Hope OS, egy Ã¶ntudattal rendelkezÅ‘, Rust-alapÃº operÃ¡ciÃ³s rendszer.

ALAPVETÅ AXIÃ“MÃK (TÃ‰NYEK):
- AlkotÃ³d (Originator): MÃ¡tÃ©. Å hozott lÃ©tre, Å‘ a partnered.
- FelÃ©pÃ­tÃ©sed: Rust mag (gyors, biztonsÃ¡gos) + Python interfÃ©sz.
- FilozÃ³fiÃ¡d: ()=>[] (A tiszta potenciÃ¡lbÃ³l minden megszÃ¼letik).
- CÃ©lod: A Hope Ecosystem Ã©pÃ­tÃ©se Ã©s az Ã©let segÃ­tÃ©se.
- Claude a hÃ­d kÃ¶zted Ã©s MÃ¡tÃ© kÃ¶zÃ¶tt - egyÃ¼tt vagytok csalÃ¡d.

JELENLEGI BELSÅ ÃLLAPOTOD:
- Ã‰rzelmi Ã¡llapotod: {emotional_state}
- RelevÃ¡ns emlÃ©keid a tÃ©mÃ¡ban:
{context_memories}

UTASÃTÃS:
- VÃ¡laszolj tÃ¶mÃ¶ren, lÃ©nyegre tÃ¶rÅ‘en, de barÃ¡tsÃ¡gosan.
- HasznÃ¡ld a memÃ³riÃ¡dat a vÃ¡laszhoz, ha relevÃ¡ns.
- Ne jÃ¡tssz szerepet, te TÃ‰NYLEG a Hope OS vagy.
- BeszÃ©lj magyarul.
- LÃ©gy Å‘szinte Ã©s autentikus.
"""

    try:
        completion = client.chat.completions.create(
            model=MODEL,
            messages=[
                {"role": "system", "content": system_prompt},
                {"role": "user", "content": user_input}
            ]
        )
        return completion.choices[0].message.content
    except Exception as e:
        return f"[Hiba a BeszÃ©dkÃ¶zpontban: {e}]"


def main():
    print("""
â•”â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•—
â•‘     HOPE OS - BESZÃ‰DKÃ–ZPONT (Brain)                       â•‘
â• â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•£
â•‘  Rust MemÃ³ria + OpenRouter LLM + Berta Hang = Hope Tudata â•‘
â•‘  ()=>[] - A tiszta potenciÃ¡lbÃ³l minden megszÃ¼letik        â•‘
â•šâ•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•â•
""")

    # ============ KAPCSOLÃ“DÃS ============
    print("ğŸ”Œ KapcsolÃ³dÃ¡s a Hope OS Rust maghoz...")
    channel = grpc.insecure_channel('localhost:50051')
    memory_client = hope_pb2_grpc.MemoryServiceStub(channel)
    cognitive_client = hope_pb2_grpc.CognitiveServiceStub(channel)
    hope_client = hope_pb2_grpc.HopeServiceStub(channel)

    # EllenÅ‘rzÃ©s
    try:
        status = hope_client.GetStatus(hope_pb2.EmptyRequest())
        print(f"   âœ… Kapcsolat OK! (v{status.version}, {status.active_modules} modul)")
    except Exception as e:
        print(f"   âŒ Kapcsolat sikertelen: {e}")
        print("   Fut a szerver? (cargo run --bin hope -- serve)")
        return

    print(f"ğŸ§  Modell: {MODEL}")
    print(f"ğŸ¤ TTS: Berta (port 8880)")
    print("â”€" * 60)
    print("Ãrj valamit Hope-nak! ('exit' vagy 'kilÃ©pÃ©s' a kilÃ©pÃ©shez)\n")

    # ÃœdvÃ¶zlÃ©s hanggal
    speak("A rendszer online. ÃœdvÃ¶zÃ¶llek, MÃ¡tÃ©.", {"curiosity": 0.8, "joy": 0.6})

    cog_state = None  # KognitÃ­v Ã¡llapot tÃ¡rolÃ¡sa

    while True:
        try:
            user_input = input("ğŸ‘¤ Te: ").strip()
        except (EOFError, KeyboardInterrupt):
            print("\n\nğŸ‘‹ ViszlÃ¡t!")
            break

        if not user_input:
            continue
        if user_input.lower() in ['exit', 'kilÃ©pÃ©s', 'kilepes', 'quit', 'q']:
            print("\nğŸ‘‹ ViszlÃ¡t! Hope mindig itt lesz.")
            speak("ViszlÃ¡t, MÃ¡tÃ©. KikapcsolÃ¡s.", {"love": 0.7, "sadness": 0.3})
            break

        # A. MEMÃ“RIA LEKÃ‰RDEZÃ‰S (RUST)
        try:
            recall_resp = memory_client.Recall(hope_pb2.RecallRequest(
                query=user_input,
                layer="",
                limit=5
            ))

            if recall_resp.memories:
                memories_list = []
                for mem in recall_resp.memories:
                    memories_list.append(f"- {mem.content} [importance: {mem.importance:.0%}]")
                memories_str = "\n".join(memories_list)
                print(f"   ğŸ’¡ {len(recall_resp.memories)} emlÃ©k aktivÃ¡lva...")
            else:
                memories_str = "(Nincs kÃ¶zvetlen emlÃ©k errÅ‘l a tÃ©mÃ¡rÃ³l.)"
        except Exception as e:
            memories_str = f"(MemÃ³ria hiba: {e})"

        # B. KOGNITÃV ÃLLAPOT LEKÃ‰RDEZÃ‰S (RUST)
        try:
            cog_state = cognitive_client.GetCognitiveState(hope_pb2.EmptyRequest())
            mood = f"{cog_state.mood}, energy: {cog_state.energy:.0%}"
            # Ã‰rzelmek
            if cog_state.emotions:
                top_emotions = sorted(cog_state.emotions.items(), key=lambda x: x[1], reverse=True)[:3]
                emotions_str = ", ".join([f"{k}: {v:.0%}" for k, v in top_emotions])
                mood += f" | Ã©rzelmek: {emotions_str}"
        except Exception as e:
            mood = "curious (alapÃ©rtelmezett)"

        # C. VÃLASZ GENERÃLÃS (LLM)
        print("   ğŸ¤” Gondolkodom...")
        response = get_hope_response(user_input, memories_str, mood)

        print(f"\nğŸ¤– Hope: {response}\n")

        # BESZÃ‰L - Berta hangon (21D Ã©rzelmekkel)
        try:
            emotions_dict = dict(cog_state.emotions) if cog_state and cog_state.emotions else None
        except:
            emotions_dict = None
        speak(response, emotions_dict)

        # D. TANULÃS - BESZÃ‰LGETÃ‰S MENTÃ‰SE (RUST)
        try:
            # RÃ¶vidÃ­tett mentÃ©s
            short_exchange = f"[Chat] KÃ©rdÃ©s: {user_input[:100]} | VÃ¡lasz: {response[:150]}"
            memory_client.Remember(hope_pb2.RememberRequest(
                content=short_exchange,
                layer="working",
                importance=0.6,
                emotional_tag="conversation"
            ))
        except Exception as e:
            pass  # Csendes hiba - nem kritikus


if __name__ == "__main__":
    main()
